README.md
# Gaphas
[![Build state](https://github.com/gaphor/gaphas/workflows/build/badge.svg)](https://github.com/gaphor/gaphas/actions)
[![Maintainability](https://api.codeclimate.com/v1/badges/e9837cc647b72119fd11/maintainability)](https://codeclimate.com/github/gaphor/gaphas/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/e9837cc647b72119fd11/test_coverage)](https://codeclimate.com/github/gaphor/gaphas/test_coverage)
![Docs build state](https://readthedocs.org/projects/gaphas/badge/?version=latest)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)
[![standard-readme compliant](https://img.shields.io/badge/readme%20style-standard-brightgreen.svg?style=flat)](https://github.com/RichardLitt/standard-readme)
[![Gitter](https://img.shields.io/gitter/room/nwjs/nw.js.svg)](https://gitter.im/Gaphor/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![All Contributors](https://img.shields.io/badge/all_contributors-9-orange.svg?style=flat-square)](#contributors)

> Gaphas is the diagramming widget library for Python.

![Gaphas Demo](docs/images/gaphas-demo.gif)

Gaphas is a library that provides the user interface component (widget) for drawing diagrams. Diagrams can be drawn to screen and then easily exported to a variety of formats, including SVG and PDF. Want to build an app with chart-like diagrams? Then Gaphas is for you! Use this library to build a tree, network, flowchart, or other diagrams.

This library is currently being used by [Gaphor](https://github.com/gaphor/gaphor) for UML drawing,
[RAFCON](https://github.com/DLR-RM/RAFCON) for state-machine based robot control, and [ASCEND](http://ascend4.org/) for solving mathematical models.

## :bookmark_tabs: Table of Contents

- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [API](#api)
- [Contributing](#contributing)
- [License](#license)

## :scroll: Background

Gaphas was built to provide the foundational diagramming portions of [Gaphor](https://github.com/gaphor/gaphor). Since Gaphor is built on GTK and Cairo, [PyGObject](https://pygobject.readthedocs.io/) provides access to the GUI toolkit and [PyCairo](https://pycairo.readthedocs.io/) to the 2D graphics library. However, there wasn't a project that abstracted these technologies to easily create a diagramming tool. Hence, Gaphas was created as a library to allow others to create a diagramming tool using GTK and Cairo.

Here is how it works:

- Items (Canvas items) can be added to a Canvas.
- The Canvas maintains the tree structure (parent-child relationships between items).
- A constraint solver is used to maintain item constraints and inter-item constraints.
- The item (and user) should not be bothered with things like bounding-box calculations.
- Very modular--e.g., handle support could be swapped in and swapped out.
- Rendering using Cairo.

The main portions of the library include:

- canvas - The main canvas class (container for Items).
- items - Objects placed on a Canvas.
- solver - A constraint solver to define the layout and connection of items.
- view - Responsible for the calculation of bounding boxes which is stored in a quadtree data structure for fast access.
- gtkview - A view to be used in GTK applications that interacts with users with tools.
- painters - The workers used to paint items.
- tools - Tools are used to handle user events (such as mouse movement and button presses).
- aspects - Provides an intermediate step between tools and items.

## :floppy_disk: Install

To install Gaphas, simply use pip:

```bash
$ pip install gaphas
```

Use of a
[virtual environment](https://virtualenv.pypa.io/en/latest/) is highly recommended.

### Development

To setup a development environment with Linux:

```bash
$ sudo apt-get install -y python3-dev python3-gi python3-gi-cairo
    gir1.2-gtk-3.0 libgirepository1.0-dev libcairo2-dev
$ pip install poetry
$ poetry install
```

## :flashlight: Usage

```python
import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk

from gaphas import Canvas, GtkView
from gaphas.examples import Box
from gaphas.painter import DefaultPainter
from gaphas.item import Line
from gaphas.segment import Segment


def create_canvas(canvas, title):
    # Setup drawing window
    view = GtkView()
    view.painter = DefaultPainter()
    view.canvas = canvas
    window = Gtk.Window()
    window.set_title(title)
    window.set_default_size(400, 400)
    win_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
    window.add(win_box)
    win_box.pack_start(view, True, True, 0)

    # Draw gaphas box
    b2 = Box(60, 60)
    b2.min_width = 40
    b2.min_height = 50
    b2.matrix.translate(170, 170)
    canvas.add(b2)

    # Draw gaphas line
    line = Line()
    line.matrix.translate(100, 60)
    canvas.add(line)
    line.handles()[1].pos = (30, 30)
    segment = Segment(line, view=None)
    segment.split_segment(0)

    window.show_all()
    window.connect("destroy", Gtk.main_quit)


c = Canvas()
create_canvas(c, "Simple Gaphas App")
Gtk.main()
```

### Overview

The `Canvas` class acts as a container for Item's . The item's parent/child relationships are maintained here, not in the `Item`.

An `Item` can have a set of handles which can be used to manipulate the item (although this is not necessary). Each item has its own coordinate system with x and y position, for example, a (0, 0) point `Item.matrix` is the transformation relative to the parent item of the `Item`, as defined in the `Canvas`.

Handles can connect to [ports](https://gaphas.readthedocs.io/en/latest/ports.html). A `Port` is a location (line or point) where a handle is allowed to connect on another
item. The process of connecting depends on the case at hand, but most often involves the creation of some sort of constraint between the Handle and the item it is connecting to.

The Canvas also contains a [constraint solver](https://gaphas.readthedocs.io/en/latest/solver.html) (from solver.py) that can be used to solve mathematical dependencies
between items (such as Handles that should be aligned). The constraint solver can also be used to keep constraints contained within the item satisfied, for example, to make sure a box maintains its rectangular shape.

A View is used to visualize a Canvas. On a View, a Tool (from tool.py) can be applied, which will handle user input like button and key presses. Painters (from painter.py) are used to do the actual drawing. This module also makes it easy to draw to other media other than a screen, such as a printer or a PDF document.

### Updating Item state

If items need updating, it sends out an update request on the Canvas (`Canvas.request_update()`). The canvas
performs an update by performing the following steps:

1. Pre-update using `Item.pre_update(context)` for each item marked for update.
2. Update the Canvas-to-Item matrices, for fast transformation of coordinates from the Canvas' to the items' coordinate      system.
3. Solve the constraints.
4. Normalize the items by setting the coordinates of the first handle to (0, 0).
5. Update the Canvas-to-Item matrices for items that have been changed by normalization.
6. Post-update using `Item.post_update(context)` for each item marked for update, including items that have been
   marked during the constraint solving step.

Gaphas attempts to do as much updating as possible in the {pre|post}_update() methods, since they are called when the application is not handling user input.

The context contains a CairoContext. This can be used, for example, to calculate the dimensions of text. One thing to keep in mind is that updating is done from the Canvas. Items should not update sub-items. After the update steps are complete, the Item should be ready to be drawn.

### Constraint solving

Constraint solving is one of the big features of this library. The Solver is able to mathematically solve these constraint rules that are applied to an item or between items. Constraints are applied to items through variables owned by
the item. An example of applying a constraint to an item is that Element items use constraints to maintain their rectangular shape. An example of applying constraints between items is to apply a constraint between a line and a box in
order to connect them.

Constraints that apply to one item are pretty straight forward, as all variables live in the same coordinate system of the item. The variables, like the Handle's x and y coordinate can simply be put in a constraint.

When two items are connected to each other and constraints are created, a problem shows up: variables live in separate coordinate systems. In order to overcome this problem, a Projection (from solver.py) has been defined. With a
Projection instance, a variable can be "projected" on another coordinate system. In this case, the Canvas' coordinate system is used when two items are connected to each other.

### Drawing

Drawing is done by the View. All items marked for redraw, the items that have been updated, will be drawn in the order in which they reside in the Canvas. The order starts with the first root item, then its children, then second root item, etc.

The view context passed to the Items draw() method has the following properties:

- view - The view we're drawing to.
- cairo - The CairoContext to draw to.
- selected - True if the item is actually selected in the view.
- focused - True if the item has the focus
- hovered - True if the mouse pointer if over the item. Only the top-most item is marked as hovered.
- dropzone - The item is marked as the drop zone. When this happens then an item is dragged over the item, and if it is     dropped, it will become a child of this item.

The View automatically calculates the bounding box for the item, based on the items drawn in the draw (context) function (this is only done when really necessary, e.g., after an update of the item). The bounding box is in viewport coordinates.

The actual drawing is done by Painters (painter.py). A series of Painters have been defined: one for handles, one for items, etc.

### Tools

Behaviour is added to the Canvas (view) by tools. Tools can be chained together in order to provide more complex behaviour.

To make it easy, a DefaultTool has been defined which is a ToolChain instance with the tools added as follows:

- ToolChain - Delegates to a set of individual tools and keeps track of which tool has grabbed the focus. This normally     happens when the user presses a mouse button. Once this happens, the tool requests a "grab" and all events, like motion   or button release, are sent directly to the focused tool.

- HoverTool - Makes the item under the mouse button the "hovered item." When such an item is drawn, its context.hovered_item flag will be set to True.

- HandleTool - Allows for handles to be dragged around and focuses the item when its handle is clicked on.

- ItemTool - Selects items and enables dragging items around.

- TextEditTool - A demo tool that features a text edit pop-up.

- RubberbandTool - Invoked when the mouse button is pressed on a section of the view where no items or handles are present. It allows the user to select items using a "rubber band" selection box.

### Interaction

Tools handle Interaction with the Canvas view (visual component). Although the default tools do a fair amount of work, in most cases you'll desire to create some custom connection behaviour. In order to implement these, HandleTool provides hooks including connect, disconnect, and glue.

One of the challenges you'll likely face is what to do when an item is removed from the Canvas and there are other items (lines) connected to it. Gaphas provides a solution to this by providing a disconnect handler to the handle instance once it is connected. A function can be assigned to this disconnect handler, which is then called when the item that is connected to is removed from the Canvas.

### Undo

Gaphas has a simple built-in system for registering changes in its classes and notifying the application. This code resides in state.py.

There is also a "reverter" framework in place. This system is notified when objects change their state, and the framework will figure out the reverse operation that has to be applied in order to undo the operation.

## :mag: API

The API can be separated into a [Model-View-Controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) with these three parts:

1. The Model, including the canvas and items
2. The View, called view
3. The Controller, called tools

### Canvas and Items

#### Class: `gaphas.canvas.Canvas`

The `Canvas` is a container for items.

```python
canvas = Canvas()
```

#### Class: `gaphas.item.Item`

Base class (or interface) for items on a `Canvas`.

```python
item = Item()
```

##### Properties:

- `matrix`: The item's transformation matrix
- `canvas`: The canvas, which owns an item
- `constraints`: list of item constraints, automatically registered when the item is added to a canvas; may be extended in subclasses

#### Class: `gaphas.connector.Handle`

Handles are used to support modifications of Items.

If the handle is connected to an item, the `connected_to` property should refer to the item. A `disconnect` handler should
be provided that handles the required disconnect behaviour, for example, cleaning up the constraints and `connected_to`.

- pos (`gaphas.connector.Position`): The position of the item, default value is (0, 0).
- strength (int): The strength of the handle to use in the constraint solver; default value is NORMAL, which is 20.
- connectable (bool): Makes the handle connectable to other items; default value is False.
- movable (bool): Makes the handle moveable; default value is True.

  ```python
  handle = Handle((10, 10), connectable=True)
  ```

#### Class: `gaphas.connector.LinePort`

The Line Port is part of an item that provides a line between two handles.

- start (`gaphas.connector.Position`): The start position of the line.
- end (`gaphas.connector.Position`): The end position of the line.

```python
p1, p2 = (0.0, 0.0), (100.0, 100.0)
port = LinePort(p1, p2)
```

#### Class: `gaphas.connector.PointPort`

The Point Port connects the handle to an item using a port at the location of the handle.

```python
h = Handle((10, 10))
port = PointPort(h.pos)
```

#### Class: `gaphas.solver.Solver`

A Solver solves constraints.

```python
a, b, c = Variable(1.0), Variable(2.0), Variable(3.0)
solver = Solver()
c_eq = EquationConstraint(lambda a,b: a+b, a=a, b=b)
solver.add_constraint(c_eq)
```

#### Class: `gaphas.constraint.EqualsConstraint`

Make 'a' and 'b' equal.

```python
a, b = Variable(1.0), Variable(2.0)
eq = EqualsConstraint(a, b)
eq.solve_for(a)
```

#### Class: `gaphas.constraint.LessThanConstraint`

Ensure one variable stays smaller than another.

```python
a, b = Variable(3.0), Variable(2.0)
lt = LessThanConstraint(smaller=a, bigger=b)
lt.solve_for(a)
```

#### Class: `gaphas.constraint.CenterConstraint`

Ensures a Variable is kept between two other variables.

```python
a, b, center = Variable(1.0), Variable(3.0), Variable()
eq = CenterConstraint(a, b, center)
eq.solve_for(a)
```

#### Class: `gaphas.constraint.EquationConstraint`

Solve a linear equation.

```python
a, b, c = Variable(), Variable(4), Variable(5)
cons = EquationConstraint(lambda a, b, c: a + b - c, a=a, b=b, c=c)
cons.solve_for(a)
```

#### Class: `gaphas.constraint.BalanceConstraint`

Keeps three variables in line, maintaining a specific ratio.

```python
a, b, c = Variable(2.0), Variable(3.0), Variable(2.3, WEAK)
bc = BalanceConstraint(band=(a,b), v=c)
c.value = 2.4
```

#### Class: `gaphas.constraint.LineConstraint`

Solves the equation where a line is connected to a line or side at a specific point.

```python
line = (Variable(0), Variable(0)), (Variable(30), Variable(20))
point = (Variable(15), Variable(4))
lc = LineConstraint(line=line, point=point)
```

### View

#### Class: `gaphas.view.View`

View class for `gaphas.canvas.Canvas` objects.

```python
canvas = Canvas()
view = View(canvas=canvas)
```

#### Class: `gaphas.view.GtkView`

GTK+ widget for rendering a `gaphas.canvas.Canvas` to a screen.

```python
canvas = Canvas()
win = Gtk.Window()
view = GtkView(canvas=canvas)
win.add(view)
```

#### Class: `gaphas.painter.PainterChain`

Chain up a set of painters.

```python
view.painter = (
    PainterChain()
    .append(FreeHandPainter(ItemPainter()))
    .append(HandlePainter())
    .append(FocusedItemPainter())
    .append(ToolPainter())
)
```

#### Class: `gaphas.painter.DrawContext`

Special context for drawing the item. It contains a cairo context and properties like selected and focused.

- **kwargs: Optional cairo properties for a context.

```python
DrawContext(
    painter=self,
    cairo=cairo,
    selected=(item in view.selected_items),
    focused=(item is view.focused_item),
    hovered=(item is view.hovered_item),
    dropzone=(item is view.dropzone_item),
)
```

#### Class: `gaphas.painter.ItemPainter`

Painter to draw an item.

```python
svgview = View(view.canvas)
svgview.painter = ItemPainter()
```

#### Class: `gaphas.painter.CairoBoundingBoxContext`

It is used to intercept `stroke()`, `fill()`, and other context operations so that the bounding box of the item involved can be calculated.

- cairo (cairo.Context): The cairo context to intercept.

```python
cairo = CairoBoundingBoxContext(cairo)
```

#### Class: `gaphas.painter.BoundingBoxPainter`

A type of ItemPainter which is used to calculate the bounding boxes (in canvas coordinates) for the items.

```python
view.bounding_box_painter = BoundingBoxPainter(ItemPainter())
```

#### Class: `gaphas.painter.HandlePainter`

Draw handles of items that are marked as selected in the view.

#### Class: `gaphas.painter.ToolPainter`

Allows the Tool defined on a view to conduct drawing.

#### Class: `gaphas.painter.FocusedItemPainter`

Used to draw on top of all the other layers for the focused item.

### Tools

Interacting with the Canvas is done through tools. Tools tell _what_ has to be done (like moving). To make an element move aspects are defined. Aspects tell how the behaviour has to be performed.

#### Class: `gaphas.tools.ToolChain`

Used to chain tools together. For example, chain a HoverTool, HandleTool, and SelectionTool in order to combine their functionality in to a new tool.

- view (`gaphas.view.View`): The view to use for the tool chain.

```python
(ToolChain(view)
.append(HoverTool())
.append(ConnectHandleTool())
.append(PanTool())
.append(ZoomTool())
.append(ItemTool())
.append(TextEditTool())
.append(RubberbandTool())
)
```

#### Class: `gaphas.tools.HoverTool`

Makes the item under the mouse cursor the hovered item.

- view (`gaphas.view.View`): The view to use for the tool; default is None.

#### Class: `gaphas.tools.ItemTool`

Does selection and dragging of items.

- view (`gaphas.view.View`): The view to use for the tool; default is None.

#### Class: `gaphas.tools.HandleTool`

Tool to move handles around.

- view (`gaphas.view.View`): The view to use for the tool; default is None.

#### Class: `gaphas.tools.RubberbandTool`

Allows the user to drag a "rubber band" for selecting items in an area.

- view (`gaphas.view.View`): The view to use for the tool; default is None.

#### Class: `gaphas.tools.PanTool`

Captures drag events with the middle mouse button and uses them to translate the Canvas within the view.

- view (`gaphas.view.View`): The view to use for the tool; default is None.

#### Class: `gaphas.tools.ZoomTool`

Tool for zooming using two different user inputs:

1. Ctrl + middle-mouse dragging in the up and down direction
2. Ctrl + mouse-wheel

- view (`gaphas.view.View`): The view to use for the tool; default is None.

#### Class: `gaphas.tools.PlacementTool`

Tool for placing items on the Canvas.

- view (`gaphas.view.View`): The view to use for the tool.
- factory (factory object): A Canvas item factory for creating new items.
- handle_tool (`gaphas.tools.HandleTool`): The handle tool to use.
- handle_index (int): The index of the handle to be used by the handle tool.

```python
def on_clicked(button):
    view.tool.grab(PlacementTool(view, factory(view, MyLine), HandleTool(), 1))
```

#### Class: `gaphas.aspects.ItemFinder`

Find an item on the Canvas.

- view (`gaphas.view.View`): The view to use in order to search for an item.

#### Class: `gaphas.aspects.ItemSelection`

Manages selection and unselection of items.

- item (`gaphas.item.Item`): The item to set as focused or unfocused.
- view (`gaphas.view.View`): The view to focus or unfocus on.

#### Class: `gaphas.aspects.ItemInMotion`

Manages motion of an item.

- item (`gaphas.item.Item`): The item to move.
- view (`gaphas.view.View`): The view to to use for move coordinates.

NEWS
2.1.1
-----
- Allow for custom update context for Canvas
- Render (GTK) background for the view
- Allow FreeHandPainter to be a child of BoundingBoxPainter

2.1.0
-----
- Fix issue where port could not be updated when reconnecting two items
- Documentation fixes
- BoundingBoxPainter now takes an Item painter as its argument
  (composition over inheritance)

2.0.1
-----
- Fix issue where undo events for handle movements were not recorded

2.0.0
-----
- Drop support for Python 2.7

1.1.2
-----
- Fix issue with placing popovers in the new

1.1.1
-----
- Fix issue where views are not rendered

1.1.0
-----
- Remove dependency on simplegeneric and decorator
- Speed up rendering by using a back-buffer
- Deprecate Handle.x and Handle.y, use Handle.pos instead

1.0.0
-----
- Change license from LGPL 2.0 to Apache 2.0
- Port to Python 3 with support for 3.5, 3.6, and 3.7
- Python < 2.7 is no longer supported
- Port from GTK+ 2 and PyGTK to GTK+ 3 with PyGObject
- Migrate tests from Nose to PyTest with Tox
- Reorganize project files
- Overhaul the README
- Add contributing guide and code of conduct
- Adopt Black as code formatter
- Add Continuous Integration with Travis and Read The Docs

0.7.2
-----
- Fix bug in calculating bounding box for rotated text
- Few minor updates

0.7.1
-----
- Views no longer lose reference to canvas on unrealize()
- bug fix in log message

0.7.0
-----
- Painters are bound to a specific view, like tools
- Introduced aspects for finding items and handles
- New feature: Guides, for aligning elements
- Free hand drawing style

0.6.0
-----
- Handlers are no longer called recursively for observed methods/properties.
- removed enable_dispatching() and disable_dispatching() calls.
- Made aspect code simpler.
- Moved disconnect code from tool to aspect, as stated in Aspect's docstring.
- Fixed issues in connections.
- Lot's of fixes and testing has been done on the undo mechanism.

0.5.0
-----
- Split tools in tools and aspects, separating the _what_ from the _how_.
  For this, a dependency to the simplegeneric module is introduced.
- Renamed VariablePoint to Position.
- Handle is no longer inheriting from VariablePoint/Position.
- Handle connections are no longer registered on the handle, but are
  maintained in the Canvas instance. This makes for much easier querying
  (e.g. which elements are attached to some element).
  Added a Table class to support this functionality.
- Added a timeout property on the @async decorator. The method is invoked
  the amunt of milliseconds after the first invocation.

0.4.1
------
- Call Item._set_canvas after matrix update
- Verify if value changed before marking variable as dirty.

0.4.0
------
- allow to define connectable parts of item's (ports feature)
- implemented default connection tool (thanks to ports feature)
- line segment tool implemented (code taken from gaphor)
- implemented Item.constraint method to simplify item's constraint
  creation
- The canvas (-view) is no longer tied to the (0, 0) position. Scrolling can
  be done quite fluidly with the new PanTool implementation.
- API changes
  - use positions instead of "x, y" pairs in all method calls


demo.py
#!/usr/bin/env python
"""
A simple demo app.

It sports a small canvas and some trivial operations:

 - Add a line/box
 - Zoom in/out
 - Split a line segment
 - Delete focused item
 - Record state changes
 - Play back state changes (= undo !) With visual updates
 - Exports to SVG and PNG

"""

import gi

gi.require_version("Gtk", "3.0")

import math

import cairo
from gi.repository import Gtk

from gaphas import Canvas, GtkView, View
from gaphas import state
from gaphas.examples import Box, PortoBox, Text, FatLine, Circle
from gaphas.freehand import FreeHandPainter
from gaphas.item import Line
from gaphas.painter import (
    PainterChain,
    ItemPainter,
    HandlePainter,
    FocusedItemPainter,
    ToolPainter,
    BoundingBoxPainter,
)
from gaphas.segment import Segment
from gaphas.tool import PlacementTool, HandleTool
from gaphas.util import text_extents, text_underline

# painter.DEBUG_DRAW_BOUNDING_BOX = True

# Ensure data gets picked well:
import gaphas.picklers

# Global undo list
undo_list = []


def undo_handler(event):
    global undo_list
    undo_list.append(event)


def factory(view, cls):
    """
    Simple canvas item factory.
    """

    def wrapper():
        item = cls()
        view.canvas.add(item)
        return item

    return wrapper


class MyBox(Box):
    """Box with an example connection protocol.
    """


class MyLine(Line):
    """Line with experimental connection protocol.
    """

    def __init__(self):
        super(MyLine, self).__init__()
        self.fuzziness = 2

    def draw_head(self, context):
        cr = context.cairo
        cr.move_to(0, 0)
        cr.line_to(10, 10)
        cr.stroke()
        # Start point for the line to the next handle
        cr.move_to(0, 0)

    def draw_tail(self, context):
        cr = context.cairo
        cr.line_to(0, 0)
        cr.line_to(10, 10)
        cr.stroke()


class MyText(Text):
    """
    Text with experimental connection protocol.
    """

    def draw(self, context):
        Text.draw(self, context)
        cr = context.cairo
        w, h = text_extents(cr, self.text, multiline=self.multiline)
        cr.rectangle(0, 0, w, h)
        cr.set_source_rgba(0.3, 0.3, 1.0, 0.6)
        cr.stroke()


class UnderlineText(Text):
    def draw(self, context):
        cr = context.cairo
        text_underline(cr, 0, 0, "Some text(y)")


def create_window(canvas, title, zoom=1.0):
    view = GtkView()
    view.painter = (
        PainterChain()
        .append(FreeHandPainter(ItemPainter()))
        .append(HandlePainter())
        .append(FocusedItemPainter())
        .append(ToolPainter())
    )
    view.bounding_box_painter = FreeHandPainter(BoundingBoxPainter())
    w = Gtk.Window()
    w.set_title(title)
    w.set_default_size(400, 120)
    h = Gtk.HBox()
    w.add(h)

    # VBox contains buttons that can be used to manipulate the canvas:
    v = Gtk.VBox()
    v.set_property("border-width", 3)
    v.set_property("spacing", 2)
    f = Gtk.Frame()
    f.set_property("border-width", 1)
    f.add(v)
    h.pack_start(f, False, True, 0)

    v.add(Gtk.Label.new("Item placement:"))

    b = Gtk.Button.new_with_label("Add box")

    def on_clicked(button, view):
        # view.window.set_cursor(Gdk.Cursor.new(Gdk.CursorType.CROSSHAIR))
        view.tool.grab(PlacementTool(view, factory(view, MyBox), HandleTool(), 2))

    b.connect("clicked", on_clicked, view)
    v.add(b)

    b = Gtk.Button.new_with_label("Add line")

    def on_clicked(button):
        view.tool.grab(PlacementTool(view, factory(view, MyLine), HandleTool(), 1))

    b.connect("clicked", on_clicked)
    v.add(b)

    v.add(Gtk.Label.new("Zooming:"))

    b = Gtk.Button.new_with_label("Zoom in")

    def on_clicked(button):
        view.zoom(1.2)

    b.connect("clicked", on_clicked)
    v.add(b)

    b = Gtk.Button.new_with_label("Zoom out")

    def on_clicked(button):
        view.zoom(1 / 1.2)

    b.connect("clicked", on_clicked)
    v.add(b)

    v.add(Gtk.Label.new("Misc:"))

    b = Gtk.Button.new_with_label("Split line")

    def on_clicked(button):
        if isinstance(view.focused_item, Line):
            segment = Segment(view.focused_item, view)
            segment.split_segment(0)
            view.queue_draw_item(view.focused_item)

    b.connect("clicked", on_clicked)
    v.add(b)

    b = Gtk.Button.new_with_label("Delete focused")

    def on_clicked(button):
        if view.focused_item:
            canvas.remove(view.focused_item)

    b.connect("clicked", on_clicked)
    v.add(b)

    v.add(Gtk.Label.new("State:"))
    b = Gtk.ToggleButton.new_with_label("Record")

    def on_toggled(button):
        global undo_list
        if button.get_active():
            print("start recording")
            del undo_list[:]
            state.subscribers.add(undo_handler)
        else:
            print("stop recording")
            state.subscribers.remove(undo_handler)

    b.connect("toggled", on_toggled)
    v.add(b)

    b = Gtk.Button.new_with_label("Play back")

    def on_clicked(self):
        global undo_list
        apply_me = list(undo_list)
        del undo_list[:]
        print("Actions on the undo stack:", len(apply_me))
        apply_me.reverse()
        saveapply = state.saveapply
        for event in apply_me:
            print("Undo: invoking", event)
            saveapply(*event)
            print("New undo stack size:", len(undo_list))
            # Visualize each event:
            # while Gtk.events_pending():
            #    Gtk.main_iteration()

    b.connect("clicked", on_clicked)
    v.add(b)

    v.add(Gtk.Label.new("Export:"))

    b = Gtk.Button.new_with_label("Write demo.png")

    def on_clicked(button):
        svgview = View(view.canvas)
        svgview.painter = ItemPainter()

        # Update bounding boxes with a temporaly CairoContext
        # (used for stuff like calculating font metrics)
        tmpsurface = cairo.ImageSurface(cairo.FORMAT_ARGB32, 0, 0)
        tmpcr = cairo.Context(tmpsurface)
        svgview.update_bounding_box(tmpcr)
        tmpcr.show_page()
        tmpsurface.flush()

        w, h = svgview.bounding_box.width, svgview.bounding_box.height
        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, int(w), int(h))
        cr = cairo.Context(surface)
        svgview.matrix.translate(-svgview.bounding_box.x, -svgview.bounding_box.y)
        cr.save()
        svgview.paint(cr)

        cr.restore()
        cr.show_page()
        surface.write_to_png("demo.png")

    b.connect("clicked", on_clicked)
    v.add(b)

    b = Gtk.Button.new_with_label("Write demo.svg")

    def on_clicked(button):
        svgview = View(view.canvas)
        svgview.painter = ItemPainter()

        # Update bounding boxes with a temporaly CairoContext
        # (used for stuff like calculating font metrics)
        tmpsurface = cairo.ImageSurface(cairo.FORMAT_ARGB32, 0, 0)
        tmpcr = cairo.Context(tmpsurface)
        svgview.update_bounding_box(tmpcr)
        tmpcr.show_page()
        tmpsurface.flush()

        w, h = svgview.bounding_box.width, svgview.bounding_box.height
        surface = cairo.SVGSurface("demo.svg", w, h)
        cr = cairo.Context(surface)
        svgview.matrix.translate(-svgview.bounding_box.x, -svgview.bounding_box.y)
        svgview.paint(cr)
        cr.show_page()
        surface.flush()
        surface.finish()

    b.connect("clicked", on_clicked)
    v.add(b)

    b = Gtk.Button.new_with_label("Dump QTree")

    def on_clicked(button, li):
        view._qtree.dump()

    b.connect("clicked", on_clicked, [0])
    v.add(b)

    b = Gtk.Button.new_with_label("Pickle (save)")

    def on_clicked(button, li):
        f = open("demo.pickled", "wb")
        try:
            import pickle

            pickle.dump(view.canvas, f)
        finally:
            f.close()

    b.connect("clicked", on_clicked, [0])
    v.add(b)

    b = Gtk.Button.new_with_label("Unpickle (load)")

    def on_clicked(button, li):
        f = open("demo.pickled", "rb")
        try:
            import pickle

            canvas = pickle.load(f)
            canvas.update_now()
        finally:
            f.close()
        create_window(canvas, "Unpickled diagram")

    b.connect("clicked", on_clicked, [0])
    v.add(b)

    b = Gtk.Button.new_with_label("Unpickle (in place)")

    def on_clicked(button, li):
        f = open("demo.pickled", "rb")
        try:
            import pickle

            canvas = pickle.load(f)
        finally:
            f.close()
        # [i.request_update() for i in canvas.get_all_items()]
        canvas.update_now()
        view.canvas = canvas

    b.connect("clicked", on_clicked, [0])
    v.add(b)

    b = Gtk.Button.new_with_label("Reattach (in place)")

    def on_clicked(button, li):
        view.canvas = None
        view.canvas = canvas

    b.connect("clicked", on_clicked, [0])
    v.add(b)

    # Add the actual View:

    view.canvas = canvas
    view.zoom(zoom)
    view.set_size_request(150, 120)
    s = Gtk.ScrolledWindow.new()
    s.set_hexpand(True)
    s.add(view)
    h.add(s)

    w.show_all()

    w.connect("destroy", Gtk.main_quit)

    def handle_changed(view, item, what):
        print(what, "changed: ", item)

    view.connect("focus-changed", handle_changed, "focus")
    view.connect("hover-changed", handle_changed, "hover")
    view.connect("selection-changed", handle_changed, "selection")


def create_canvas(c=None):
    if not c:
        c = Canvas()
    b = MyBox()
    b.min_width = 20
    b.min_height = 30
    b.matrix = (1.0, 0.0, 0.0, 1, 20, 20)
    b.width = b.height = 40
    c.add(b)

    bb = Box()
    bb.matrix = (1.0, 0.0, 0.0, 1, 10, 10)
    c.add(bb, parent=b)

    fl = FatLine()
    fl.height = 50
    fl.matrix.translate(100, 100)
    c.add(fl)

    circle = Circle()
    h1, h2 = circle.handles()
    circle.radius = 20
    circle.matrix.translate(50, 100)
    c.add(circle)

    # AJM: extra boxes:
    bb = Box()
    bb.matrix.rotate(math.pi / 1.567)
    c.add(bb, parent=b)
    # for i in xrange(10):
    #     bb = Box()
    #     print('box', bb)
    #     bb.matrix.rotate(math.pi/4.0 * i / 10.0)
    #     c.add(bb, parent=b)

    b = PortoBox(60, 60)
    b.min_width = 40
    b.min_height = 50
    b.matrix.translate(55, 55)
    c.add(b)

    t = UnderlineText()
    t.matrix.translate(70, 30)
    c.add(t)

    t = MyText("Single line")
    t.matrix.translate(70, 70)
    c.add(t)

    l = MyLine()
    c.add(l)
    l.handles()[1].pos = (30, 30)
    segment = Segment(l, view=None)
    segment.split_segment(0, 3)
    l.matrix.translate(30, 60)
    l.orthogonal = True

    off_y = 0
    for align_x in (-1, 0, 1):
        for align_y in (-1, 0, 1):
            t = MyText(
                "Aligned text %d/%d" % (align_x, align_y),
                align_x=align_x,
                align_y=align_y,
            )
            t.matrix.translate(120, 200 + off_y)
            off_y += 30
            c.add(t)

    t = MyText("Multiple\nlines", multiline=True)
    t.matrix.translate(70, 100)
    c.add(t)

    return c


def main():
    ##
    ## State handling (a.k.a. undo handlers)
    ##

    # First, activate the revert handler:
    state.observers.add(state.revert_handler)

    def print_handler(event):
        print("event:", event)

    c = Canvas()

    create_window(c, "View created before")

    create_canvas(c)

    # state.subscribers.add(print_handler)

    ##
    ## Start the main application
    ##

    create_window(c, "View created after")

    Gtk.main()


if __name__ == "__main__":
    import sys

    if "-p" in sys.argv:
        print("Profiling...")
        import hotshot, hotshot.stats

        prof = hotshot.Profile("demo-gaphas.prof")
        prof.runcall(main)
        prof.close()
        stats = hotshot.stats.load("demo-gaphas.prof")
        stats.strip_dirs()
        stats.sort_stats("time", "calls")
        stats.print_stats(20)
    else:
        main()

# vim: sw=4:et:
